### PR TITLE
ENH: Add axis argument to random.permutation and random.shuffle

### PR DESCRIPTION
**Pull request by kianasun**
The `random.permutation` function now can only shuffle the first axis of a multi-dimensional array. I add an argument for the function and allow it to shuffle along a given axis.

### CODE REVIEW

File: numpy/random/generator.pyx
Comment (wkschwartz) [inline numpy/random/generator.pyx]: Should there be some input validation in case `axis` is a slice or something weird?

**Review by wkschwartz — COMMENTED** [review]
Needs documentation and possibly input validation/error handling for the new argument.

File: numpy/random/generator.pyx
Comment (kianasun) [reply numpy/random/generator.pyx]: Thanks for your reply! I added codes for error handling and updated the documentation :)

Comment (wkschwartz) [reply numpy/random/generator.pyx]: I think what you want is `numpy.core.mulitarray.normalize_axis_index`, documented here https://github.com/numpy/numpy/blob/5a08e209863038b411bc506893093c5a4b8377c9/numpy/add_newdocs.py#L6739

Comment (kianasun) [reply numpy/random/generator.pyx]: I didn't know this function! Thanks! Updated it.

Comment (wkschwartz) [reply numpy/random/generator.pyx]: `normalize_axis_index` is all you need. You can remove the `isinstance` check and raising the error.

```python console
>>> import numpy
>>> numpy.__version__
'1.16.0'
>>> numpy.core.multiarray.normalize_axis_index(slice(1, 4, 2), 3)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: an integer is required (got type slice)
>>> numpy.core.multiarray.normalize_axis_index((slice(1, 4, 2),), 3)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: an integer is required (got type tuple)
>>> numpy.core.multiarray.normalize_axis_index((1, 2, None), 3)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: an integer is required (got type tuple)
>>> numpy.core.multiarray.normalize_axis_index(-1, 3)
2
>>> numpy.core.multiarray.normalize_axis_index(1, 3)
1
```

File: numpy/random/tests/test_random.py
Comment (wkschwartz) [inline numpy/random/tests/test_random.py]: Also do the same test with `axis=-4` to make sure you get the same thing as you do with `axis=1`.

File: numpy/random/generator.pyx
Comment (kianasun) [reply numpy/random/generator.pyx]: I didn't remove `isinstance` because the input of `normalize_axis_index` is an `int`. But if it can also correctly raise the error, I can remove it.

File: numpy/random/tests/test_random.py
Comment (kianasun) [reply numpy/random/tests/test_random.py]: You are correct, we should also test negative axis. The tested array is a 2d array, maybe `axis=-1`, since it should be `-ndim <= axis < ndim` for `normalize_axis_index`?

Comment (wkschwartz) [reply numpy/random/tests/test_random.py]: You're right. I conflated axis and column.

Comment (kianasun) [reply numpy/random/tests/test_random.py]: Updated! :)

File: numpy/random/mtrand.pyx
Comment (wkschwartz) [inline numpy/random/mtrand.pyx]:
```suggestion
```

**Comment by wkschwartz** [comment]
Once that print statement is removed, I think this is ready for review by core maintainers. @kianasun, the [docs](https://docs.scipy.org/doc/numpy/dev/gitwash/development_workflow.html#asking-for-your-changes-to-be-merged-with-the-main-repo) say that the next step is for you to e-mail the [NumPy mailing list](https://scipy.org/scipylib/mailing-lists.html) to request a code review.

File: numpy/random/mtrand.pyx
Comment (kianasun) [reply numpy/random/mtrand.pyx]: Sorry. Forgot to remove.

**Comment by kianasun** [comment]
@wkschwartz OK. I will send an email after all checks are done. Thank you!

**Comment by wkschwartz** [comment]
Should this PR be extended to address #11583, and there by make the two functions have more similar signatures?

File: numpy/random/mtrand.pyx
Comment (bashtage) [inline numpy/random/mtrand.pyx]: This file is frozen and cannot be changed except for issues relating to compilation, not features or bugs.  These changes need to be removed.

**Review by bashtage — CHANGES_REQUESTED** [review]
You can't add any features to mtrand.pyx.

**Comment by bashtage** [comment]
It would be nice to mirror the api in shuffle as well since permutation and shuffle are closely related. Shuffle could be implemented using two calls to `numpy.swapaxes`

File: numpy/random/mtrand.pyx
Comment (wkschwartz) [reply numpy/random/mtrand.pyx]: https://github.com/numpy/numpy/pull/13838#pullrequestreview-254540116 suggests doing a documentation-cleaning pass on `numpy.random` before 1.17 is cut. Perhaps that would be a good time to add a comment at the top of `mtrand.pyx` to forewarn future contributors from adding features to the legacy code.

Comment (bashtage) [reply numpy/random/mtrand.pyx]: Agree. It should be clear that NEP-19 has frozen this file.

Comment (kianasun) [reply numpy/random/mtrand.pyx]: Oh, I see. If I cannot add features to `mtrand.pyx`, can you please refer which file should I add features to?

Comment (bashtage) [reply numpy/random/mtrand.pyx]: New features can be implemented for Generator in generator.pyx.  They cannot be added to RandomState (in mtrand.pyx).

Comment (kianasun) [reply numpy/random/mtrand.pyx]: Ok, then I will only keep the changes in `generator.pyx`, and also add similar feature to `shuffle`.

**Comment by kianasun** [comment]
Hi @bashtage , I have updated the code :)

File: numpy/random/generator.pyx
Comment (bashtage) [inline numpy/random/generator.pyx]: This needs some thought.  Perpahs you should raise TypeError here since it isn't really possible to use axis with things that aren't ndarray or subclasses.

File: numpy/random/generator.pyx
Comment (bashtage) [inline numpy/random/generator.pyx]: Add an example of a multidimensional shuffle using axis>0 (or axis=-1, or both).

File: numpy/random/generator.pyx
Comment (bashtage) [inline numpy/random/generator.pyx]: Example below would also be a good idea.

File: numpy/random/generator.pyx
Comment (bashtage) [inline numpy/random/generator.pyx]: Should axis!=0 raise ValueError here?

**Comment by bashtage** [comment]
Some small issues but this needs to go to the mailing list since the API is a bit tricky since shuffle has 3 distinct paths/uses and only one needs axis.

File: numpy/random/generator.pyx
Comment (wkschwartz) [reply numpy/random/generator.pyx]: `normalize_axis_index` should have already done the right thing by this line:

```python console
>>> import numpy
>>> numpy.core.multiarray.normalize_axis_index(0, 1)
0
>>> numpy.core.multiarray.normalize_axis_index(1, 1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
numpy.AxisError: axis 1 is out of bounds for array of dimension 1
>>> numpy.core.multiarray.normalize_axis_index(-1, 1)
0
>>> numpy.core.multiarray.normalize_axis_index(-2, 1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
numpy.AxisError: axis -2 is out of bounds for array of dimension 1
>>> issubclass(numpy.AxisError, ValueError)
True
```

File: numpy/random/generator.pyx
Comment (wkschwartz) [reply numpy/random/generator.pyx]: Maybe there should also be a `numpy.AxisError` in the "fast path" above if `axis` is neither 0 nor -1.

File: numpy/random/generator.pyx
Comment (bashtage) [reply numpy/random/generator.pyx]: Ok, great. Having seen both together permutation looks a lot less controversial than shuffle since the first thing permutation does is call asarray.

File: numpy/random/tests/test_generator_mt19937.py
Comment (wkschwartz) [inline numpy/random/tests/test_generator_mt19937.py]: The tests immediately above this cover a variety of input types. Should the `axis` argument be paired in combination with them?

File: numpy/random/tests/test_generator_mt19937.py
Comment (wkschwartz) [inline numpy/random/tests/test_generator_mt19937.py]: Could be a separate method `test_permutation_custom_axis` like https://github.com/numpy/numpy/blob/3deeea9522f5c512d52b3fe6d418d99b338eaece/numpy/random/tests/test_generator_mt19937.py#L586-L597

File: numpy/random/generator.pyx
Comment (wkschwartz) [reply numpy/random/generator.pyx]: Should there be tests of exceptions like https://github.com/numpy/numpy/blob/3deeea9522f5c512d52b3fe6d418d99b338eaece/numpy/random/tests/test_generator_mt19937.py#L598-L615

File: numpy/random/generator.pyx
Comment (kianasun) [reply numpy/random/generator.pyx]: @bashtage Do you mean raising `TypeError` for `# Untyped path`? Because `x` has already been confirmed as a `ndarray` in the other two cases.

File: numpy/random/tests/test_generator_mt19937.py
Comment (kianasun) [reply numpy/random/tests/test_generator_mt19937.py]: I am not sure. If combine them, we might need to test errors for different input types. For example `lambda x: np.array([])` which has one dimension, if we set `axis=1`, do we need to assert `AxisError`?

Comment (eric-wieser) [reply numpy/random/tests/test_generator_mt19937.py]: Yes, `axis=1` should raise `AxisError` on a 1d array

Comment (kianasun) [reply numpy/random/tests/test_generator_mt19937.py]: In that case, will it be better to combine them since we need to raise errors for some but not all types?

**Comment by kianasun** [comment]
Hi @bashtage @wkschwartz, I have updated the code as recommended (except a few questions I asked). By the way, do I need to change anything to pass the check `Shippable`? I am not sure why it is failed. Thanks.

**Comment by bashtage** [comment]
@kianasun You should propose this enhancement on the NumPy mailing list.

**Comment by kianasun** [comment]
Hi @bashtage, I sent an email to Numpy-discussion last week.

File: numpy/random/generator.pyx
Comment (kianasun) [reply numpy/random/generator.pyx]: @eric-wieser In this case, we can make sure it is an `ndarray`, do we still need to check `if axis != 0`? I am just wondering when will it fail.

Comment (eric-wieser) [reply numpy/random/generator.pyx]: @kianasun: I am talking about the `# Untyped path` which this comment chain is in response to.

Comment (kianasun) [reply numpy/random/generator.pyx]: @eric-wieser I see. I will add this for `Untyped path`

File: numpy/random/generator.pyx
Comment (kianasun) [reply numpy/random/generator.pyx]: @eric-wieser Updated.

File: numpy/random/generator.pyx
Comment (wkschwartz) [inline numpy/random/generator.pyx]: `axis != -1` is always true because `0 <= normalize_axis_index(axis, np.ndim(x)) < np.ndim(x)` is always true. See @eric-wieser's [comment above](https://github.com/numpy/numpy/pull/13829/files#r324054975)

Comment (kianasun) [reply numpy/random/generator.pyx]: @wkschwartz oh, yes.. you are right.

**Comment by mattip** [comment]
The mailing list discussion was positive. I will merge this soon if there are no more comments.

File: numpy/random/generator.pyx
Comment (eric-wieser) [inline numpy/random/generator.pyx]: This doesn't make any sense to me - the message it prints out will just be wrong if a.ndim isn't 1.

File: numpy/random/generator.pyx
Comment (eric-wieser) [inline numpy/random/generator.pyx]: For simplicity, can you just call `normalize_axis_index ` unconditionally at the start of the function?

**Comment by eric-wieser** [comment]
This should be ENH not MAINT, and needs a release note.

File: numpy/random/generator.pyx
Comment (kianasun) [reply numpy/random/generator.pyx]: @eric-wieser It seems for untyped path, the function only shuffle on `0` axis?

File: numpy/random/generator.pyx
Comment (wkschwartz) [reply numpy/random/generator.pyx]: Should there be a test case with axis != 0 that hits the untyped path?

File: numpy/random/generator.pyx
Comment (kianasun) [reply numpy/random/generator.pyx]: @eric-wieser But you are right, the error message is not correct. Should we raise another error?

File: numpy/random/generator.pyx
Comment (kianasun) [reply numpy/random/generator.pyx]: @eric-wieser sure.

File: numpy/random/generator.pyx
Comment (eric-wieser) [reply numpy/random/generator.pyx]: I think on this path, you should use:
```
if axis != 0:
    x = np.swapaxes(x, ...)
```

That way, if `swapaxes` is supported on a duck-array, it will just work. If it isn't supported, then the duckarray will be able to report itself that its not supported.

By adding the `!=0` check, you ensure that duckarrays don't need to implement swapaxes to support axis=0.

File: numpy/random/generator.pyx
Comment (eric-wieser) [reply numpy/random/generator.pyx]: Note that you'll need to use `np.ndim(x)` instead of `x.ndim`

File: numpy/random/generator.pyx
Comment (kianasun) [reply numpy/random/generator.pyx]: @eric-wieser I see. If `swapaxes` works, we need to add another `swapaxes` after the block`with self.lock ...` ?

Comment (eric-wieser) [reply numpy/random/generator.pyx]: Correct

File: numpy/random/generator.pyx
Comment (eric-wieser) [inline numpy/random/generator.pyx]: Because `shuffle` shuffles in-place, this line is pointless - `x` is not used after this point

File: numpy/random/generator.pyx
Comment (eric-wieser) [inline numpy/random/generator.pyx]: I've realized this doesn't work so well if `x` is a list.

@shoyer: Any suggestions for allowing `__array_function__` overrides here without letting a copy occur?

@kianasun: Sorry for the indecision. Perhaps instead we should revert this to:
```
if axis != 0:
    raise NotImplementedError("Axis argument is only supported on ndarray objects")
```
then we can solve this properly later.

Comment (kianasun) [reply numpy/random/generator.pyx]: @eric-wieser Sure.

File: numpy/random/generator.pyx
Comment (kianasun) [reply numpy/random/generator.pyx]: @wkschwartz I added one for `random.shuffle`.

**Review by eric-wieser — APPROVED** [review]

File: numpy/random/generator.pyx
Comment (wkschwartz) [inline numpy/random/generator.pyx]: Should this be documented here and/or for permutation?

Comment (eric-wieser) [reply numpy/random/generator.pyx]: This doesn't apply to `permutation`, but would indeed be worth mentioning for `shuffle`

Comment (kianasun) [reply numpy/random/generator.pyx]: @wkschwartz @eric-wieser Added one sentence in the doc of `axis`.

**Comment by mattip** [comment]
Thanks @kianasun and reviewers

**Comment by WarrenWeckesser** [comment]
FYI: There's a small bug in the new code.  I proposed a fix in #14662.

**Pull request merged**
